### ISSUE TITLE
no filter, but tests being discarded?

### ISSUE THREAD

**Issue by aryairani** [typelevel/scalacheck#172]
Any idea what could be going on here?   I've tried adding verbose to see what was being discarded, but I didn't observe any change in output.  (P.S. I'm stuck on scalacheck `1.12.2` due to https://github.com/scalaz/scalaz/pull/954)

```
[info] SetExtractorProps
[info] 
[error]   x EmptySet extractor
[error]  Gave up after only 26 passed tests. 132 tests were discarded. (SetExtractors.scala:56)
[info] 
[error]   x NonEmptySet extractor
[error]  Gave up after only 32 passed tests. 162 tests were discarded. (SetExtractors.scala:56)
```

``` scala
import org.specs2.{Specification, ScalaCheck}

object EmptySet {
  def unapply[A](s: Set[A]): Boolean = s.isEmpty
}

object NonEmptySet {
  def unapply[A](s: Set[A]): Option[(A, Set[A])] = if (s.isEmpty) Option.empty else Option((s.head, s.tail))
}

object SetExtractors {
  sealed trait Result
  case object Empty extends Result
  case object NonEmpty extends Result
  case object NoMatch extends Result

  def emptySetExtractor[A]: Set[A] ⇒ Result = {
    case EmptySet() ⇒ Empty
    case _ ⇒ NoMatch
  }

  def nonemptySetExtractor[A]: Set[A] ⇒ Result = {
    case NonEmptySet(one, more) ⇒ NonEmpty
    case _ ⇒ NoMatch
  }
}

class SetExtractorProps extends Specification with ScalaCheck {
//  override implicit val defaultParameters = (new Parameters).verbose
  override def is = s2"""

  EmptySet extractor    $empty
  NonEmptySet extractor $nonempty
  """

  import SetExtractors._
  def empty = prop { (s: Set[Int]) ⇒ ((emptySetExtractor(s) == Empty) mustEqual s.isEmpty) }
  def nonempty = prop { (s: Set[Int]) ⇒ (nonemptySetExtractor(s) == NonEmpty) mustEqual !s.isEmpty }
}
```

**Comment by rickynils** [comment]
I suspect you're running into the problem reported in #89

**Comment by aishfenton** [comment]
@rickynils what is the suggested way to make a container of exactly size N? (I'm running into #89 since I'm creating a 2-dim array, and I can't have it being jagged).

**Comment by rickynils** [comment]
@aishfenton What kind of containers are you generating? For arrays and lists, `Gen.containerOfN` should work perfectly fine. How is #89 affecting you?

**Comment by aishfenton** [comment]
Ahh I think it maybe due to shrinking (which I'm only now understanding).

I am trying to generate a square array
 
```scala
val inner = Gen.containerOfN[Array, Double](50, Gen.choose(0.0, 1.0))
val outer = Gen.containerOfN[Array, Array[Double]](50, inner)
```

But I'm getting different sizes for inner and outer when I use it with forall(outer) (in Scalatest). But don't see any problems sampling from outer directly (in the console). So I assume shrinking is causing the problem?

**Comment by aishfenton** [comment]
Ok. Found this works. Required bit I was missing was *suchThat*.

```scala
val inner = Gen.containerOfN[Array, Double](50, Gen.choose(0.0, 1.0)).suchThat(_.size == 50)
val outer = Gen.containerOfN[Array, Array[Double]](50, inner).suchThat(_size == 50)
```

But it doesn't work if I use a for comprehension?

```scala
for { 
  inner <- Gen.containerOfN[Array, Double](50, Gen.choose(0.0, 1.0)).suchThat(_.size == 50);
  outer <- Gen.containerOfN[Array, Array[Double]](50, inner).suchThat(_.size == 50)
} yield outer
```

Probably totally obvious to those in the know, but I'm still a little mystified.

**Comment by ashawley** [comment]
I believe the test failures for discarded values were fixed by #575 and #714.

Running similar tests with the generators will pass now.

```scala
val inner: Gen[Array[Double]] = Gen.containerOfN[Array, Double](50, Gen.choose(0.0, 1.0)).suchThat(_.size == 50)
val outer: Gen[Array[Array[Double]]] = Gen.containerOfN[Array, Array[Double]](50, inner).suchThat(_.size == 50)

val outer2: Gen[Array[Array[Double]]] = for {
  inner <- Gen.containerOfN[Array, Double](50, Gen.choose(0.0, 1.0)).suchThat(_.size == 50);
  outer <- Gen.containerOfN[Array, Array[Double]](50, inner).suchThat(_.size == 50)
} yield outer

property("outer") = Prop.forAll(outer) { (arrs) =>
  arrs.size == 50 && arrs.forall(_.size == 50)
}

property("outer2") = Prop.forAll(outer) { (arrs) =>
  arrs.size == 50 && arrs.forall(_.size == 50)
}
```
